app/models/subsidiary.rb

```diff
@@ -1,2 +1,25 @@
 class Subsidiary < ApplicationRecord
+  validates :name, :address, :cnpj, presence: true
+  validates :cnpj, :name, :token, uniqueness: true
+
+  validate :cnpj_must_be_valid
+
+  before_create :generate_token
+
+  def cnpj_must_be_valid
+    return if cnpj.blank? || CNPJ.valid?(cnpj)
+
+    errors.add :cnpj, :invalid
+  end
+
+  private
+
+  def generate_token
+    generated_token = ''
+    loop do
+      generated_token = SecureRandom.alphanumeric(6).upcase
+      break if Subsidiary.find_by(token: generated_token).blank?
+    end
+    self.token = generated_token
```

Comment:
Acho que a variável poderia estar dentro do escopo do método e expor ela pra fora, porque aí você pode fazer algo como.

```suggestion
    self.token = loop do
      generated_token = SecureRandom.alphanumeric(6).upcase
      break generated_token  if Subsidiary.exists?(token: generated_token)
    end
```

Ao retornar a variável com `break` você recebe esse valor de fora então pode ficar mais tranquilo.

Troquei o find_by(...).blank? por `exists?` porque aí fica menos verboso o código mas nesse caso aí nem faz tanta diferença. Faria em outros casos porque o `exists?` não tem cache ele sempre faz a query, o que nesse caso faz diferença zero.

Reply:
Valeu, Henrique! Não sabia que o break retornava tb, legal!